File: vespajlib/src/main/java/com/yahoo/slime/Value.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.slime;

import java.io.ByteArrayOutputStream;
import java.util.function.Consumer;

/**
 * Common implementation for all value types.
 * All default behavior is here, so specific types only
 * need override their actually useful parts
 *
 * @author havardpe
 */
abstract class Value implements Cursor {

    static final String emptyString = "";
    static final byte[] emptyData = new byte[0];

    public final boolean valid() { return this != NixValue.invalid(); }

    public final void ifValid(Consumer<Inspector> consumer) {
        if (valid()) consumer.accept(this);
    }

    public int children() { return 0; }
    public int entries() { return 0; }
    public int fields() { return 0; }

    public boolean asBool() { return false; }
    public long asLong() { return 0; }
    public double asDouble() { return 0.0; }
    public String asString() { return emptyString; }
    public byte[] asUtf8() { return emptyData; }
    public byte[] asData() { return emptyData; }

    public void traverse(ArrayTraverser at) {}
    public void traverse(ObjectSymbolTraverser ot) {}
    public void traverse(ObjectTraverser ot) {}

    public Value entry(int idx) { return NixValue.invalid(); }
    public Value field(String name) { return NixValue.invalid(); }
    public Value field(int sym) { return NixValue.invalid(); }

    protected Cursor addLeaf(Value value) { return NixValue.invalid(); }
    public Cursor addArray() { return NixValue.invalid(); }
    public Cursor addObject() { return NixValue.invalid(); }

    public final Cursor addNix() { return addLeaf(NixValue.instance()); }
    public final Cursor addBool(boolean bit) { return addLeaf(BoolValue.instance(bit)); }
    public Cursor addLong(long l) { return NixValue.invalid(); }
    public Cursor addDouble(double d) { return NixValue.invalid(); }
    public final Cursor addString(String str) { return addLeaf(StringValue.create(str)); }
    public final Cursor addString(byte[] utf8) { return addLeaf(Utf8Value.create(utf8)); }
    public final Cursor addData(byte[] data) { return addLeaf(DataValue.create(data)); }

    protected Cursor setLeaf(int sym, Value value) { return NixValue.invalid(); }
    public Cursor setArray(int sym) { return NixValue.invalid(); }
    public Cursor setObject(int sym) { return NixValue.invalid(); }

    public final Cursor setNix(int sym) { return setLeaf(sym, NixValue.instance()); }
    public final Cursor setBool(int sym, boolean bit) { return setLeaf(sym, BoolValue.instance(bit)); }
    public final Cursor setLong(int sym, long l) { return setLeaf(sym, new LongValue(l)); }
    public final Cursor setDouble(int sym, double d) { return setLeaf(sym, new DoubleValue(d)); }
    public final Cursor setString(int sym, String str) { return setLeaf(sym, StringValue.create(str)); }
    public final Cursor setString(int sym, byte[] utf8) { return setLeaf(sym, Utf8Value.create(utf8)); }
    public final Cursor setData(int sym, byte[] data) { return setLeaf(sym, DataValue.create(data)); }

    protected Cursor setLeaf(String name, Value value) { return NixValue.invalid(); }
    public Cursor setArray(String name) { return NixValue.invalid(); }
    public Cursor setObject(String name) { return NixValue.invalid(); }

    public final Cursor setNix(String name) { return setLeaf(name, NixValue.instance()); }
    public final Cursor setBool(String name, boolean bit) { return setLeaf(name, BoolValue.instance(bit)); }
    public final Cursor setLong(String name, long l) { return setLeaf(name, new LongValue(l)); }
    public final Cursor setDouble(String name, double d) { return setLeaf(name, new DoubleValue(d)); }
    public final Cursor setString(String name, String str) { return setLeaf(name, StringValue.create(str)); }
    public final Cursor setString(String name, byte[] utf8) { return setLeaf(name, Utf8Value.create(utf8)); }
    public final Cursor setData(String name, byte[] data) { return setLeaf(name, DataValue.create(data)); }

    public final String toString() {
        try {
            // should produce non-compact json, but we need compact
            // json for slime summaries until we have a more generic
            // json rendering pipeline in place.
            ByteArrayOutputStream a = new ByteArrayOutputStream();
            new JsonFormat(true).encode(a, this);
            byte[] utf8 = a.toByteArray();
            return Utf8Codec.decode(utf8, 0, utf8.length);
        } catch (Exception e) {
            return "null";
        }
    }
}

File: vespalib/src/vespa/vespalib/data/slime/value.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "value.h"
#include "nix_value.h"
#include "resolved_symbol.h"
#include "empty_value_factory.h"
#include "basic_value_factory.h"
#include "external_data_value_factory.h"
#include <vespa/vespalib/data/simple_buffer.h>
#include "json_format.h"

namespace vespalib::slime {

bool
Value::valid() const {
    return (this != NixValue::invalid());
}

Type
Value::type() const {
    return NIX::instance;
}

size_t
Value::children() const {
    return 0;
}

size_t
Value::entries() const {
    return 0;
}

size_t
Value::fields() const {
    return 0;
}

// default NOPs for leaf values
Cursor &
Value::addLeaf(const ValueFactory &) { return *NixValue::invalid(); }
Cursor &
Value::setLeaf(Symbol, const ValueFactory &) { return *NixValue::invalid(); }
Cursor &
Value::setLeaf(Memory, const ValueFactory &) { return *NixValue::invalid(); }

// default values for accessors
bool
Value::asBool() const { return false; }
int64_t
Value::asLong() const { return 0; }
double
Value::asDouble() const { return 0.0; }
Memory
Value::asString() const { return Memory(); }
Memory
Value::asData() const { return Memory(); }

Cursor &
Value::operator[](size_t) const { return *NixValue::invalid(); }
Cursor &
Value::operator[](Symbol) const { return *NixValue::invalid(); }
Cursor &
Value::operator[](Memory) const { return *NixValue::invalid(); }

// default NOPs for traversal
void
Value::traverse(ArrayTraverser &) const {}
void
Value::traverse(ObjectSymbolTraverser &) const {}
void
Value::traverse(ObjectTraverser &) const {}

// generate string representation
vespalib::string
Value::toString() const
{
    SimpleBuffer buf;
    slime::JsonFormat::encode(*this, buf, false);
    return buf.get().make_string();
}

// 7 x add
Cursor &
Value::addNix() { return addLeaf(NixValueFactory()); }
Cursor &
Value::addBool(bool bit) { return addLeaf(BoolValueFactory(bit)); }
Cursor &
Value::addLong(int64_t l) { return addLeaf(LongValueFactory(l)); }
Cursor &
Value::addDouble(double d) { return addLeaf(DoubleValueFactory(d)); }
Cursor &
Value::addString(Memory str) { return addLeaf(StringValueFactory(str)); }
Cursor &
Value::addData(Memory data) { return addLeaf(DataValueFactory(data)); }
Cursor &
Value::addData(ExternalMemory::UP data) { return addLeaf(ExternalDataValueFactory(std::move(data))); }

// 7 x set (with numeric symbol id)
Cursor &
Value::setNix(Symbol sym) { return setLeaf(sym, NixValueFactory()); }
Cursor &
Value::setBool(Symbol sym, bool bit) { return setLeaf(sym, BoolValueFactory(bit)); }
Cursor &
Value::setLong(Symbol sym, int64_t l) { return setLeaf(sym, LongValueFactory(l)); }
Cursor &
Value::setDouble(Symbol sym, double d) { return setLeaf(sym, DoubleValueFactory(d)); }
Cursor &
Value::setString(Symbol sym, Memory str) { return setLeaf(sym, StringValueFactory(str)); }
Cursor &
Value::setData(Symbol sym, Memory data) { return setLeaf(sym, DataValueFactory(data)); }
Cursor &
Value::setData(Symbol sym, ExternalMemory::UP data) { return setLeaf(sym, ExternalDataValueFactory(std::move(data))); }

// 7 x set (with symbol name)
Cursor &
Value::setNix(Memory name) { return setLeaf(name, NixValueFactory()); }
Cursor &
Value::setBool(Memory name, bool bit) { return setLeaf(name, BoolValueFactory(bit)); }
Cursor &
Value::setLong(Memory name, int64_t l) { return setLeaf(name, LongValueFactory(l)); }
Cursor &
Value::setDouble(Memory name, double d) { return setLeaf(name, DoubleValueFactory(d)); }
Cursor &
Value::setString(Memory name, Memory str) { return setLeaf(name, StringValueFactory(str)); }
Cursor &
Value::setData(Memory name, Memory data) { return setLeaf(name, DataValueFactory(data)); }
Cursor &
Value::setData(Memory name, ExternalMemory::UP data) { return setLeaf(name, ExternalDataValueFactory(std::move(data))); }

// nop defaults for array/objects
Cursor &
Value::addArray(size_t) { return *NixValue::invalid(); }
Cursor &
Value::addObject() { return *NixValue::invalid(); }

Cursor &
Value::setArray(Symbol, size_t) { return *NixValue::invalid(); }
Cursor &
Value::setObject(Symbol) { return *NixValue::invalid(); }
Cursor &
Value::setArray(Memory, size_t) { return *NixValue::invalid(); }
Cursor &
Value::setObject(Memory) { return *NixValue::invalid(); }

Symbol
Value::resolve(Memory) { return Symbol(); }

} // namespace vespalib::slime
